skimage/restoration/_denoise.py

```diff
@@ -401,19 +402,25 @@ def _denoise_tv_chambolle_nd(image, weight=0.1, eps=2.e-4, n_iter_max=200):
             out = image + d
         else:
             out = image
-        E = (d ** 2).sum()
+        E = xp.sum(d * d)
 
         # g stores the gradients of out along each axis
         # e.g. g[0] is the first order finite difference along axis 0
         slices_g = [slice(None), ] * (ndim + 1)
         for ax in range(ndim):
             slices_g[ax+1] = slice(0, -1)
             slices_g[0] = ax
-            g[tuple(slices_g)] = np.diff(out, axis=ax)
+
+            # first order difference along the axis (np.diff not in array API)
+            _at = functools.partial(utils.slice_at_axis, axis=ax)
+            g[tuple(slices_g)] = (
+                out[_at(slice(1, None))] - out[_at(slice(0, -1))]
+            )
+
             slices_g[ax+1] = slice(None)
 
-        norm = np.sqrt((g ** 2).sum(axis=0))[np.newaxis, ...]
-        E += weight * norm.sum()
+        norm = xp.expand_dims(xp.sqrt(xp.sum(g * g, axis=0)), axis=0)
```

Comment:
I think this is reasonable alternative. Although, there are 3 nested `xp` calls, so I think it is a little less readable.